phpseclib/Crypt/Random.php

```diff
@@ -260,6 +260,44 @@ function crypt_random_string($length)
     }
 }
 
+if (!function_exists('phpseclib_safe_serialize')) {
+    /**
+     * Safely serialize variables
+     *
+     * If a class has a private __sleep() method it'll give a fatal error on PHP 5.2 and earlier.
+     * PHP 5.3 will emit a warning.
+     *
+     * @param mixed $arr
+     * @param array $refs optional
+     * @access public
+     */
+    function phpseclib_safe_serialize($arr, $refs = array())
+    {
+        if (is_object($arr)) {
+            return '';
+        }
+        if (!is_array($arr)) {
+            return serialize($arr);
+        }
+        $safearr = array();
+        foreach (array_keys($arr) as $key) {
+            if (is_object($arr[$key])) {
+                continue;
+            }
+            // without this recursive arrays (eg. $a = []; $a[] = &$a;) would result in an
+            // "Allowed memory size of ... bytes exhausted" Fatal error
+            foreach ($refs as $ref) {
+                if ($ref === $arr[$key]) {
```

Comment:
```php
$GLOBALS['a'] = &$GLOBALS;

phpseclib_safe_serialize($GLOBALS, 'GLOBALS');
```

That causes the same "Fatal error: Allowed memory size" error. I mean, on one hand, people ought not be doing that, but then again, people ought not be making `__sleep` private either.

In light of this, what about this?:

https://github.com/terrafrost/phpseclib/commit/086bd845fe236213bb9f6860dd6c2c879974fb77